File: gherkin/dotnet/Gherkin/CucumberMessages/Types/RuleChild.cs
using System.Runtime.Serialization;

namespace Gherkin.CucumberMessages.Types
{
    public class RuleChild
    {
        [DataMember(Name = "background")]
        public Background Background { get; set; }

        [DataMember(Name = "scenario")]
        public Scenario Scenario { get; set; }

        public RuleChild()
        {
        }

        public RuleChild(Background background, Scenario scenario)
        {
            Background = background;
            Scenario = scenario;
        }
    }
}

File: gherkin/dart/lib/src/ast/RuleChild.dart
import 'package:gherkin/ast.dart';
import 'package:gherkin/core.dart';

class RuleChild
    implements INullSafetyObject
{
  static const empty = _InvalidRuleChild();

  final Background background;

  final Scenario scenario;

  const RuleChild.fromBackground(this.background)
      : scenario = Scenario.empty;

  const RuleChild.fromScenario(this.scenario)
      : background = Background.empty;

  @override
  bool get isEmpty => false;

  @override
  bool get isNotEmpty => !isEmpty;
}

class _InvalidRuleChild extends RuleChild
{
  const _InvalidRuleChild()
      : super.fromBackground(Background.empty);

  @override
  bool get isEmpty => true;
}
